File: leetcode/leetcode53.java
// beats over 67% of submission
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums.length == 0 || nums == null){
            return 0;
        }
        if(nums.length == 1){
            return nums[0];
        }
        
        // sum[i]: maximum sum of subarray ends at index i
        int[] sum = new int[nums.length];
        sum[0] = nums[0];
        int max = sum[0];
        
        for(int i = 1; i < nums.length; i++){
            sum[i] = Math.max(sum[i-1] + nums[i], nums[i]);
            if(sum[i] > max){
                max = sum[i];
            }
        }
        
        return max;
    }
}

// another idea
// how to get the max sum of subarray which ends at specific number
// e.g. 2, -3, 4, -1 for ending at -1, which head would we choose to start for subarray
// we already know prefix sum ends at -1 is 2
// subarray ends at 2 is 2, which means [-3, 4, -1] sum is 2 - 2 = 0
// subarray ends at -3 is -1, which means [4, -1] sum is 2 - (-1) = 3
// subarray ends at 4 is 3, which means [-1] sum is 2 - 3 = -1
// so the max sum of subarray ending at -1 is 3
// only beats over 67% of submission
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];
        
        int[] prefixSum = new int[nums.length+1];
        prefixSum[0] = 0;
        int max = Integer.MIN_VALUE;
        int min = 0;
        
        for(int i = 1; i <= nums.length; i++){
            prefixSum[i] = prefixSum[i-1] + nums[i-1];
            max = Integer.max(max, prefixSum[i] - min);
            min = Integer.min(min, prefixSum[i]);
        }
        
        return max;
    }
}

File: leetcode/leetcode53.cpp
/**
 * 這種情況下dp[i]不能定義為nums[0..i]中的最大subarr和
 * 因為狀態轉移不好寫，我們不確定最大subarr在何處，無法確保連續
 * 我們可以把dp[i]定義為以nums[i]作結的最大subarr
 * 不過這樣最後return的答案就不能是dp[n - 1]，我們需要在遍歷一次整個dp數組
 * 狀態轉移方程：
 * 遍歷到nums[i]時，要嘛nums[i]自己成為一個subarray，要嘛nums[i]跟之前的subarray concate在一起
 */
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        if(n == 1) return nums[0];
        
        vector<int> dp(n, nums[0]);
        for(int i = 1; i < n; i++) 
            dp[i] = max(nums[i], dp[i - 1] + nums[i]);
        
        int res = INT_MIN;
        for(int i = 0; i < n; i++)
            res = max(res, dp[i]);
        
        return res;
    }
};